Feature/atrayano/depends on

atrayano opened the pull request:
<!--- These lines are comments. You can delete or ignore them -->
<!--- NOTE: If your PR is trivial, feel free to delete the "Related Issue" -->
<!---       "Testing" or other sections. -->

<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

## Related Issue
<!--- This project primarily accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? --> This was requested by Arlindo. The new keywords added to the export specs automatically enable allocation of variables based on the content

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. --> 
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Trivial change (affects only documentation or cleanup)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x ] I have tested this change with a run of GEOSgcm (if non-trivial)
- [x ] I have added one of the required labels (0 diff, 0 diff trivial, 0 diff structural, non 0-diff)
- [ x] I have updated the CHANGELOG.md accordingly following the style of [Keep a Changelog](https://keepachangelog.com/en/1.0.0/#how)

tclune commented on CHANGELOG.md:
```diff
@@ -8,7 +8,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
-
+- Added support for "depends_on" and "depends_on_children" for export_specs
```
This is an important and non-trivial feature.   I think the change log needs a few sentences to indicate at least when a user might want these.

atrayano replied:
Ok, I will add these shortly

atrayano replied:
Done

tclune commented on generic/MAPL_Generic.F90:
```diff
@@ -540,12 +540,86 @@ recursive subroutine MAPL_GenericSetServices ( GC, RC )
+       ! get the export specs
+       call  MAPL_StateGetVarSpecs(meta, export=ex_specs, _RC)
+       ! allow for possibility we do not have export specs
+       if (.not. associated(ex_specs)) then
+          _RETURN(ESMF_SUCCESS)
+       end if
```
```suggestion
       _RETURN_IF(.not. associated(ex_specs))
```

atrayano commented:
The CHANGELOG has been changed. As far as testing: I did the following
1) Ran GEOS without the changes to produce a baseline results
2) Compiled GOES with the version of MAPL from this PR, but without the new keywords in any export (the no-harm test). This was 0-diff
3) Added 3 new, 2d exports (for testing purposes only), foo, bar1, and bar2, to GOCART2G. And specified DEPEDS_ON=['bar1','bar2'] for 'foo'. Then in the run method got the pointers to bar1 and bar2 and asserted that they were allocated. Similarly, asserted the those pointer were not allocated with the DEPENDS_ON being commented out for 'foo'
4) added 'DEPENDS_ON_CHILDREN=.true' for 'foo' without adding 'foo' to children. This failed as expected (this is a requirement as per Arlindo) with appropriate error message
5) Added 'foo' to some of the children of GOCART2G. This fails as expected (same requirement as above) with appropriate error message
6) Added 'foo' to all children of GOCART2G, and asserted that 'foo' for each child is allocated
